compiler/rustc_macros/src/diagnostics/diagnostic.rs

```diff
@@ -198,3 +210,31 @@ impl Mismatch {
         }
     }
 }
+
+/// Generates a `#[test]` that verifies that all referenced variables
+/// exist on this structure.
+#[cfg(debug_assertions)]
+fn generate_test(slug: &syn::Path, structure: &Structure<'_>) -> TokenStream {
+    use std::sync::atomic::{AtomicUsize, Ordering};
+    // We need to make sure that the same diagnostic slug can be used multiple times without causing an
+    // error, so just have a global counter here.
+    static COUNTER: AtomicUsize = AtomicUsize::new(0);
+    let slug = slug.get_ident().unwrap();
+    let ident = quote::format_ident!("verify_{slug}_{}", COUNTER.fetch_add(1, Ordering::Relaxed));
+    let slug = quote::format_ident!("{slug}_refs");
+    let variables: Vec<_> = structure
+        .variants()
+        .iter()
+        .flat_map(|v| v.ast().fields.iter().filter_map(|f| f.ident.as_ref().map(|i| i.to_string())))
+        .collect();
+    quote! {
+        #[cfg(test)]
+        #[test ]
```

Comment:
Tidy complains about `#[test]` in non-test files

Reply:
Maybe add a comment?